nautobot/dcim/signals.py

```diff
@@ -66,53 +67,68 @@ def handle_rackgroup_location_change(instance, created, **kwargs):
-    through child locations to find the "right" one, the best we can do is simply to null out the location.
+    through child locations to find the "right" one, the best we can do is to raise to raise a ValidationError and roll back the changes we made.
     """
-    if not created:
-        if instance.location is not None:
+    with transaction.atomic():
+        if not created:
             descendants = instance.location.descendants(include_self=True)
             content_types = instance.location.location_type.content_types.all()
             rack_groups_permitted = ContentType.objects.get_for_model(RackGroup) in content_types
             racks_permitted = ContentType.objects.get_for_model(Rack) in content_types
             power_panels_permitted = ContentType.objects.get_for_model(PowerPanel) in content_types
-        else:
-            descendants = None
-            rack_groups_permitted = False
-            racks_permitted = False
-            power_panels_permitted = False
-
-        for rackgroup in instance.children.all():
-            changed = False
-
-            if instance.location is not None:
-                if rackgroup.location is not None and rackgroup.location not in descendants:
-                    rackgroup.location = instance.location if rack_groups_permitted else None
-                    changed = True
-
-            if changed:
-                rackgroup.save()
-
-        for rack in Rack.objects.filter(group=instance):
-            changed = False
-
-            if instance.location is not None:
-                if rack.location is not None and rack.location not in descendants:
-                    rack.location = instance.location if racks_permitted else None
-                    changed = True
 
-            if changed:
-                rack.save()
-
-        for powerpanel in PowerPanel.objects.filter(rack_group=instance):
-            changed = False
-
-            if instance.location is not None:
-                if powerpanel.location is not None and powerpanel.location not in descendants:
-                    powerpanel.location = instance.location if power_panels_permitted else None
-                    changed = True
-
-            if changed:
-                powerpanel.save()
+            for rackgroup in instance.children.all():
+                changed = False
+
+                if rackgroup.location not in descendants:
+                    if rack_groups_permitted:
+                        rackgroup.location = instance.location
+                        changed = True
+                    else:
+                        raise ValidationError(
+                            {
+                                f"location {instance.location.name}": "RackGroups may not associate to locations of type "
+                                f'"{instance.location.location_type}"'
+                            }
+                        )
+
+                if changed:
+                    rackgroup.save()
```

Comment:
Could this be simplified as follows here and below?

```suggestion
                if rackgroup.location not in descendants:
                    if not rack_groups_permitted:
                        raise ValidationError(
                            {
                                f"location {instance.location.name}": "RackGroups may not associate to locations of type "
                                f'"{instance.location.location_type}"'
                            }
                        )
                    rackgroup.location = instance.location
                    rackgroup.save()
```